### PR TITLE
ComponentWillUnmount should only ever be invoked once

### PR DESCRIPTION
**Pull request by jimfb**
ComponentWillUnmount should only ever be invoked once

### CODE REVIEW

**Comment by jimfb** [comment]
@spicyj @sebmarkbage

**Comment by sebmarkbage** [comment]
Mostly because we don't have a better idea of whether it is better to just always try/catch or if we can reuse another flag or store where in the tree we left off.

File: src/renderers/shared/reconciler/ReactCompositeComponent.js [hunk 1]
```diff
@@ -132,6 +132,9 @@ var ReactCompositeComponentMixin = {
 
     // See ReactUpdates and ReactUpdateQueue.
     this._pendingCallbacks = null;
+
+    // ComponentWillUnmount shall only be called once
+    this.calledComponentWillUnmount = false;
```
Comment (sebmarkbage) [inline src/renderers/shared/reconciler/ReactCompositeComponent.js]: Use underscore for consistency.

File: src/renderers/shared/reconciler/__tests__/ReactCompositeComponent-test.js [hunk 2]
```diff
@@ -1303,4 +1303,47 @@ describe('ReactCompositeComponent', function() {
...
+
+    var setRef = (ref) => {
+      if (ref) {
+        app = ref;
+      }
+    };
+
+    expect(function() {
+      ReactDOM.render(<App ref={setRef} />, container);
+      stage = 2;
```
Comment (spicyj) [inline src/renderers/shared/reconciler/__tests__/ReactCompositeComponent-test.js]: Any reason not to pass stage as a prop to App?

**Pull request merged**